src/dataflow/src/render/reduce.rs

```diff
@@ -1212,66 +1359,114 @@ where
+                                aggr,
+                                accum,
+                                non_nulls
+                            );
+                        }
+                        _ => (),
                     }
 
                     // The finished value depends on the aggregation function in a variety of ways.
-                    let value = match (&aggr.func, agg2) {
-                        (AggregateFunc::Count, _) => Datum::Int64(agg2 as i64),
-                        (AggregateFunc::All, _) => {
+                    let value = match (&aggr.func, accum) {
+                        (AggregateFunc::Count, Accumulator::SimpleNumber { non_nulls, .. }) => {
+                            Datum::Int64(i64::cast_from(*non_nulls))
+                        }
+                        (
+                            AggregateFunc::All,
+                            Accumulator::Bool {
+                                falses,
+                                trues,
+                                total,
+                            },
+                        ) => {
                             // If any false, else if all true, else must be no false and some nulls.
-                            if agg2 > 0 {
+                            if *falses > 0 {
                                 Datum::False
-                            } else if tot == agg1 {
+                            } else if trues == total {
                                 Datum::True
                             } else {
                                 Datum::Null
                             }
                         }
-                        (AggregateFunc::Any, _) => {
+                        (
+                            AggregateFunc::Any,
+                            Accumulator::Bool {
+                                falses,
+                                trues,
+                                total,
+                            },
+                        ) => {
                             // If any true, else if all false, else must be no true and some nulls.
-                            if agg1 > 0 {
+                            if *trues > 0 {
                                 Datum::True
-                            } else if tot == agg2 {
+                            } else if falses == total {
                                 Datum::False
                             } else {
                                 Datum::Null
                             }
                         }
                         (AggregateFunc::Dummy, _) => Datum::Dummy,
                         // Below this point, anything with only nulls should be null.
-                        (_, 0) => Datum::Null,
+                        (
+                            _,
+                            Accumulator::SimpleNumber {
+                                non_nulls: 0,
+                                total: 1..=isize::MAX,
```

Comment:
this match will compile down to something reasonable right? we won't be linearly scanning through this range to see if total matches it?

Reply:
tried testing this out in compiler explorer couldn't convince it to emit useful output

Reply:
It would be great to drop a comment in here explaining, for sure. Is it something that can be match guarded with a `total > 0` instead (which strikes me as more clear, but opinions could vary).

Reply:
It is a bit at adds with the logic above, where e.g. the `Any`/`All` aggregation logic don't check to see if total is positive, and in the error cases that this catches they will just produce an output. Perhaps it might make more sense to validate the data first if we want that (i.e. `total > 0`) and then decode it?

Hey related to that: the previous version of the code factored out `total` from the "variants"; would it make sense to do that here too, using a `(Accumulator, isize)` where the second coordinate always gets a `1` and can be tested for strict positivity? It seems like (not knowing future plans for decimals) that is where this is headed, and splitting it out reduces the chances that there might be varying implementations for the different variants.

Reply:
IIUC this branch is saying "if the input was all nulls and we know there was at least some input, return a null", and the equivalent logic in the boolean case would be:

```
if trues == 0 && falses == 0 && total > 0
```

which i think is implicitly caught by the two branches in the any / all cases. I'm not sure factoring out `total` would make this more clear I'm not sure what would. I also think this pr is fine to merge as is (assuming we know that the existing match is not pathological) and I'm happy to tinker and iterate on this after that?

Reply:
Just to come to what frank said earlier about a match guard:

```
diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index 3226e1811..6d4e7baf9 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs
@@ -1439,10 +1439,10 @@ where
                             _,
                             Accumulator::SimpleNumber {
                                 non_nulls: 0,
-                                total: 1..=isize::MAX,
+                                total,
                                 ..
                             },
-                        ) => Datum::Null,
+                        ) if *total > 0 => Datum::Null,
                         // If any non-nulls, just report the aggregate.
                         (AggregateFunc::SumInt32, Accumulator::SimpleNumber { accum, .. }) => {
                             Datum::Int64(*accum as i64)
```

compiles and probably does the right thing (haven't tested) and is maybe a simple enough change to throw in?

Reply:
This definitely compiles down to something reasonable! https://godbolt.org/z/rYMdndPGf

Personally I find it clearer to avoid match guards whenever something is expressible via a pattern, but clearly empirically that opinion is not widely shared. I'll update.

Reply:
omg ofc it wasn't working when I was testing it because I was using a `fn` instead of a `pub fn`!! I see now that there's a line of text at the bottom that says "Output (0/10)" indicating 0 happy lines from the compiler and 10 angry ones :D